src/org/sage_one_sample/sageone/SageoneSigner.java

```diff
@@ -0,0 +1,130 @@
+package org.sage_one_sample.sageone;
+
+import java.net.*;
+import java.util.*;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.UnsupportedEncodingException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+public class SageoneSigner {
+  String method;
+  String url;
+  TreeMap<String, String> params;
+  String nonce;
+  String secret;
+  String token;
+
+  SageoneSigner(String requestMethod, String requestUrl, TreeMap<String, String> requestBodyParams, String requestNonce, String signingSecret, String accessToken) {
+    method = requestMethod;
+    url = requestUrl;
+    params = requestBodyParams;
+    nonce = requestNonce;
+    secret = signingSecret;
+    token = accessToken;
+  }
+
+  public String signature() {
+    String result = null;
+    try {
+      result = Base64.getEncoder().encodeToString(hmacSha1());
+    } catch (UnsupportedEncodingException | NoSuchAlgorithmException | InvalidKeyException e) {
+      //handle exceptions
+    }
+    return result;
+  }
+
+  private byte[] hmacSha1() throws UnsupportedEncodingException, NoSuchAlgorithmException, InvalidKeyException {
+    final String type = "HmacSHA1";
+    final SecretKeySpec secret = new SecretKeySpec(signingKey().getBytes(), type);
+    final Mac mac = Mac.getInstance(type);
+    mac.init(secret);
+    return mac.doFinal(signatureBaseString().getBytes());
+  }
+
+  private String signatureBaseString() {
+    String signatureBaseString = "";
+    signatureBaseString += method.toUpperCase();
+    signatureBaseString += "&";
```

Comment:
I believe every time you do a `+=` it will create a copy in memory. So instead you could chain the methods to build the string. However performance is not an issue in this so I wouldn't worry. There is also such thing as a `StringBuilder` which may ( i don't know ) be more performant

```
signatureBaseString += method.toUpperCase() + "&" +
                 percentEncode(baseUrl()) + "&" +
                 percentEncode(parameterString()) + "&"
                 percentEncode(nonce);

```